lib/bot_framework/session.rb

```diff
@@ -0,0 +1,422 @@
+        dialog_error_message: nil,
+        actions: nil
+      }
+      @auto_batch_delay = 250 unless options[:auto_batch_delay].is_a? Integer
+      @timers = Timers::Group.new
+    end
+
+    def to_recognize_context
+      {
+        message: message,
+        user_data: user_data,
+        conversation_data: conversation_data,
+        private_conversation_data: private_conversation_data,
+        localizer: localizer,
+        dialog_stack: dialog_stack,
+        preferred_locale: preferred_locale,
+        get_text: get_text,
+        nget_text: nget_text,
+        locale: preferred_locale
+      }
+    end
+
+    def dispatch(session_state, message)
+      index = 0
+      session = self
+      now = Time.now
+      middleware = @options[:middleware] || []
+
+      _next = lambda do
+        handler = middleware[index] if index < middleware.length
+        if handler
+          index += 1
+          handler(session, _next)
+        else
+          @in_middleware = false
+          @session_state[:last_acess] = now
+          done
+        end
+      end
+      session_state ||= { call_stack: [], last_acess: Time.now, version: 0.0 }
+      # Making sure that dialog is properly initialized
+      cur = cur_dialog
+      self.dialog_data = cur.state if cur
+      # Dispatch message
+      message ||= { text: '' }
+      message[:type] ||= 'message'
+      # Ensure that localized prompts are loaded
+      # TODO
+      self
+    end
+
+    def error error
+      logger.info "Session error"
+      if options[:dialog_error_message] 
+        end_conversation(options[:dialog_error_message])
+      else
+        #TODO Add localisation
+        locale= preferred_locale
+        end_conversation ("Error in conversation")
+      end
+
+      # TODO Log error
+
+    end
+    
+    def preferred_locale(locale=nil,&block)
+      if locale
+        @_locale = locale
+        @user_data['BotBuilder.Data.PreferredLocale'] = locale if @user_data
+        @localizer.load() if @localizer #TODO
+      elsif !@_locale
+        if @user_data && @user_data['BotBuilder.Data.PreferredLocale']
+          @_locale = @user_data['BotBuilder.Data.PreferredLocale']
+        elsif @message && @message[:text_locale]
+          @_locale = @message[:text_locale]
+        elsif @localizer 
+          @_locale = @localizer[:default_locale]
+        end
+      end
+      @_locale
+    end
+    
+    # Gets and formats localized text string
+    def gettext(message_id,options={})
+      #TODO
+      #stub
+    end
+
+    # Gets and formats the singular/plural form of a localized text string. 
+    def ngettext(message_id,message_id_plural,count)
+    end    
+
+    # Manually save current session state
+    def save
+      logger.info 'Session.save'
+      start_batch
+      self
+    end
+
+    def send(message,args=[])
+      args.unshift(@cur_library_name,message)
+      send_localized(args,message)
+    end
+
+    def send_localized(localization_namspace,message,args=[])
```

Comment:
Unused method argument - `args`. If it's necessary, use `_` or `_args` as an argument name to indicate that it won't be used.